annotation-server/src/medications/medications.service.ts

```diff
@@ -22,17 +23,100 @@ export class MedicationsService {
         private medicationRepository: Repository<Medication>,
     ) {}
 
-    async clearAllMedicationData(): Promise<void> {
-        await this.medicationRepository.delete({});
+    getAll(
+        options: FindManyOptions<Medication> = {
+            select: ['id', 'name', 'description', 'synonyms', 'drugclass'],
+        },
+    ): Promise<Medication[]> {
+        return this.medicationRepository.find(options);
+    }
+
+    getOne(options: FindOneOptions<Medication>): Promise<Medication> {
+        return this.medicationRepository.findOneOrFail(options);
+    }
+
+    getDetails(id: number): Promise<Medication> {
+        return this.getOne({
+            where: { id },
+            relations: [
+                'guidelines',
+                'guidelines.genePhenotype.phenotype',
+                'guidelines.genePhenotype.geneSymbol',
+            ],
+        }).catch(() => {
+            throw new NotFoundException('Medication could not be found!');
+        });
+    }
+
+    async findMatchingMedications(query = ''): Promise<Medication[]> {
+        const result = await this.medicationRepository.query(`
+        SELECT distinct id, name, description, drugclass, indication,
+        CASE 
+            WHEN name ILIKE '%${query}%' THEN 1
+            WHEN drugclass ILIKE '%${query}%' THEN 2
+            WHEN synonym ILIKE '%${query}%' THEN 3 
+            ELSE 4 
+        END as rank
+        FROM (
+            SELECT id, name, description, drugclass, indication, unnest(synonyms) synonym
+            FROM public.medication
+        ) sub
+        WHERE name ILIKE '%${query}%' OR drugclass ILIKE '%${query}%' OR synonym ILIKE '%${query}%'
```

Comment:
We also need to add a match in the description like in the version before. E. g. `Acetaminophen` should also visible if searching for `Tylenol`, which is only included in the description. This match should have the lowest priority.